lib/src/utf8.dart

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:convert';
+import 'dart:ffi';
+import 'dart:typed_data';
+
+/// [Utf8] implements conversion between Dart strings and null-terminated
+/// Utf8-encoded "char*" strings in C.
+//
+// TODO(https://github.com/dart-lang/ffi/issues/4): No need to use
+// 'asExternalTypedData' when Pointer operations are performant.
+class Utf8 extends Struct<Utf8> {
+  /// Returns the length of a null-terminated string -- the number of (one-byte)
+  /// characters before the first null byte.
+  static int strlen(Pointer<Utf8> string) {
+    final Pointer<Uint8> array = string.cast<Uint8>();
+    int count = 0x1000;
+    Uint8List nativeString = array.asExternalTypedData(count: count);
+    int i = 0;
+    while (nativeString[i] != 0) {
+      if (++i == count) {
+        count *= 2;
```

Comment:
Could we use a precise bound here?

Reply:
No, because the length of the string can be only be detected by scanning until the first NULL byte.

Reply:
But, why times 2 every time? Why not scan until null and use that as count?

Reply:
This is scanning until null.

Reply:
Ah, I was confused by `array.asExternalTypedData(count: count)` where that count is potentially 2x the string length. We need an oversized TypedData wrapper in order to not create a new wrapper every character for scanning in the first place. (I hope we can switch to the normal `Pointer<Int8>.load()` soon, that would make it easier to read this code.)